src/java.base/share/classes/java/util/stream/ForEachOps.java

```diff
@@ -499,7 +510,7 @@ else if (spliterator != null) {
             // "happens-before" completion of the associated left-most leaf task
             // of right subtree (if any, which can be this task's right sibling)
             //
-            ForEachOrderedTask<S, T> leftDescendant = completionMap.remove(this);
+            var leftDescendant = (ForEachOrderedTask<S, T>)NEXT.getAndSet(this, null);
```

Comment:
Casting the `null` is required for the resolved method descriptor to be `(ForEachOrderedTask, ForEachOrderedTask)ForEachOrderedTask` instead of `(ForEachOrderedTask, Object)ForEachOrderedTask`:
```suggestion
            var leftDescendant = (ForEachOrderedTask<S, T>) NEXT.getAndSet(this, (ForEachOrderedTask<S, T>) null);
```

Reply:
@ExE-Boss Ah, sorry, it was meant to be there. :)

Reply:
VarHandles are more optimized than MethodHandles when the signature at the call site does not match the signature of the handle. For referenced casts of arguments VarHandles avoid the `asType` conversion that would occur with method handles. This makes it much easier to write/read. 

VarHandle::withInvokeExactBehavior was introduced to deal with cases where an exact signature match is required. Sometimes this is useful to avoid performance potholes resulting from unintended primitive casts/conversions, where VarHandles otherwise resort to `asType` conversion.